snapcraft/internal/dirs.py

```diff
@@ -103,10 +103,10 @@ def setup_dirs() -> None:
         common.set_keyringsdir(os.path.join(parent_dir, "keyrings"))
         common.set_legacy_snapcraft_dir(os.path.join(snap_path, "legacy_snapcraft"))
 
-    elif sys.platform == "win32":
+    elif sys.platform == "win32" or (site.USER_BASE and topdir.startswith(site.USER_BASE + os.sep)):
         common.set_plugindir(os.path.join(topdir, "snapcraft", "plugins"))
 
-        data_dir = _find_windows_data_dir(topdir)
+        data_dir = _find_data_dir(topdir)
```

Comment:
I like the generic `find_data_dir()`.  Though I'm wondering if we can just merge the two blocks here.  If I'm looking at it correctly, the defaults are already checked in find_data_dir() because it includes a check for `sys.prefix` `[/share/snapcraft]`:
```
_DEFAULT_PLUGINDIR = os.path.join(sys.prefix, "share", "snapcraft", "plugins")
_plugindir = _DEFAULT_PLUGINDIR
_DEFAULT_SCHEMADIR = os.path.join(sys.prefix, "share", "snapcraft", "schema")
_schemadir = _DEFAULT_SCHEMADIR
_DEFAULT_EXTENSIONSDIR = os.path.join(sys.prefix, "share", "snapcraft", "extensions")
_extensionsdir = _DEFAULT_EXTENSIONSDIR
_DEFAULT_KEYRINGSDIR = os.path.join(sys.prefix, "share", "snapcraft", "keyrings")
_keyringsdir = _DEFAULT_KEYRINGSDIR
```

What if we replace `elif sys.platform == "win32" or (site.USER_BASE and topdir.startswith(site.USER_BASE + os.sep)):` with `else`, and retain the checks, i.e.:

```
    else:
        common.set_plugindir(os.path.join(topdir, "snapcraft", "plugins"))

        data_dir = _find_data_dir(topdir)
        common.set_schemadir(os.path.join(data_dir, "schema"))
        common.set_extensionsdir(os.path.join(data_dir, "extensions"))
        common.set_keyringsdir(os.path.join(data_dir, "keyrings"))

        # Make sure required data directories exist in the given locations.
        for d in [
            common.get_schemadir(),
            common.get_extensionsdir(),
            common.get_keyringsdir(),
        ]:
            if not os.path.exists(d):
                raise snapcraft.internal.errors.SnapcraftDataDirectoryMissingError()
```

Then we can ultimately do away with the defaults...?  Thoughts?

Reply:
I agree, that seems reasonable. My only concern is the plugindir, which isn't topdir/snapcraft/plugins by default. Let me look into this a little.